### DISCUSSION TITLE
Calls from HAL layer

### DISCUSSION THREAD

**Discussion by c0nwild** [elheck/Squiddrone#321] [Coding] [answered]
Regarding calls originating in HAL layer (i.e. Interrupt Request callback fired) we need some best practices.

**Comment by c0nwild** [comment]
One option that (at least) compiles is this:
```
class InterruptHandler {
 public:
  static auto HandleInterrupt() noexcept -> void {
    driver_->HandleIRQ();
  };

  explicit ComInterruptHandler(std::shared_ptr<Driver> driver_object) {
    driver_ = driver_object;
  };

 private:
  static std::shared_ptr<Driver> driver_;
};
```

**Comment by elheck** [comment, marked as answer]
Mixing a class that can be instantiated with static methods is misleading. One could come to the conclusion that you need to instantiate this class in order to use it, but you could just use the static method without ever calling the constructor. Your constructor is a glorified setter for the driver.

Because we are working with a C interrupt we cannot circumvent the statefull static class/struct which basically is a singleton. We can at least make it less painful. We can get rid of a this pointer that could be shared with many modules, which causes a lot of problems in singletons.
My proposal:

 1.   =delete default constructor
 2.   Don't implement an own constructor at all, that you cannot call any constructor at all.
 3.   make a static SetDriver method to set the driver pointer.
 4.   check if driver is nullptr in HandleInterrupt so that you can enforce that it needs to be set before using it.

A side note: This is a good practice in this case, but should not be seen as a best practice. We can only have a good practice, because we have a statefull construct (we need to set and later use the driver pointer). If we could get rid of the statefullness, we would only be using the class as a namespace for static functions and should consequently get rid of the class and should encapsulate the free static functions in just a namespace and not a class.
